eeps/eep-0079.md

```diff
@@ -0,0 +1,769 @@
+
+### Ordering and equality
+
+With addition of native records the runtime values of different types
+are now ordered as follows:
+
+```erlang
+number()
+< atom()
+< reference()
+< fun()
+< port()
+< pid()
+< tuple()
+< record()
+< map()
+< []
+< [_|_]
+< bitstring()
+```
+
+Native-record values are ordered by their fully qualified name, then
+by their visibility, then their keys, and finally by field values
+(field values are considered as maps). Equality is defined through
+equality of all the properties: names, visibility, keys and field
+values.
+
+### Reflection
+
+A new module `records` provides functionality for basic runtime
+reflection. This is a very preliminary sketch of how it may look:
+
+```erlang
+-module(records).
+-spec get_module(record()) -> module().
+-spec get_name(record()) -> atom().
+-spec is_exported(record()) -> boolean().
+-spec get_field_names(record()) -> [atom()].
+-spec get(record(), atom()) -> dynamic().
+-spec create(Module :: module(), RecordName :: atom(), FieldsMap :: #{atom() => term()}) -> record().
+-spec create(Module :: module(), RecordName :: atom(), FieldsMap :: #{atom() => term()}, Options :: #{ exported => boolean() }) -> record().
+-spec update(Src::record(), Module :: module(), RecordName :: atom(), FieldsMap :: #{atom() => term()}) -> record().
+```
+
+This part needs more thought about enforcing and bypassing visibility
+(exported and private native-record definitions).
+
+* It should provide a means to work with any native-record through
+  reflection, but bypassing visibility and opacity restrictions should
+  be explicit.
+
+The functions in the `records` module are BIFs.
+
+### Printing
+
+Here is how printed native-records will look like:
+
+```erlang
+%% native-record
+#users:user{id = 1, name = "Alice", city = "London"}
+```
+
+Printing of fields follows the field order. Whether a native-record value is
+exported is not visible through printing.
+
+### External term format
+
+External term format is extended to support serialization of native-record values.
+
+### Tooling
+
+While the compiler does not validate native-record operations at
+compile time, validation can be easily performed through simple
+linting that ensures:
+
+* The visibility of native-records is respected (a non-exported
+  native-record definition is not used outside of the defining module)
+
+* Fields are used correctly
+
+All of these checks are straightforward, and `xref` can be easily extended to handle them.
+
+### Performance characteristics
+
+While the implementation of native-records is a large scope, from the
+runtime perspective they are much closer to maps than to records
+(tuples). So, their performance characteristics should align with maps
+(with insignificant overhead for runtime validation). Additionally,
+given that native-records are more specialized versions of maps (with
+all keys being atoms), there is potential for optimizations.
```

Comment:
Is the expectation that native records will have inferior performance to tuple records?